File: src-tauri/src/commands/agent.rs
use serde::Deserialize;
use crate::ai::client::{self, ChatMessage, AiSettings};
use crate::cancel::CancelManager;
use std::path::PathBuf;
use std::fs;
use tokio::io::AsyncReadExt;

#[derive(Debug, Deserialize)]
pub struct AgentRunConfig {
    pub rect_id: String,
    pub prompt: String,
    pub existing_code: Option<String>,
    pub agent_type: String,
    // Custom agent fields
    pub command: Option<String>,
    pub args_template: Option<String>,
    pub history: Option<Vec<String>>,
}

fn work_dir(rect_id: &str) -> PathBuf {
    dirs::home_dir()
        .map(|h| h.join(".pagesprite").join("tmp").join(rect_id))
        .unwrap_or_else(|| std::env::temp_dir().join("pagesprite").join(rect_id))
}

/// Extract code from an AI response — tries ```html first, then any ``` block.
fn extract_code(text: &str) -> Option<String> {
    // Try ```html block first
    if let Some(start) = text.find("```html") {
        let after = &text[start + 7..];
        if let Some(end) = after.find("```") {
            return Some(after[..end].trim().to_string());
        }
    }
    // Try any code block
    if let Some(start) = text.find("```") {
        let after = &text[start + 3..];
        let body = if let Some(nl) = after.find('\n') {
            &after[nl + 1..]
        } else {
            after
        };
        if let Some(end) = body.find("```") {
            return Some(body[..end].trim().to_string());
        }
    }
    None
}


/// Interpolate `{dir}` and `{contextFile}` in a template string.
/// Does NOT replace `{prompt}` — the prompt is always appended as a
/// single separate argument to avoid whitespace splitting.
fn interpolate(template: &str, dir: &str, context_file: &str) -> String {
    template
        .replace("{dir}", dir)
        .replace("{contextFile}", context_file)
}

// ── Streaming agent (Rust-side AI API call) ──────────────────────────

async fn run_streaming_agent(
    config: &AgentRunConfig,
    settings: &AiSettings,
    cancel_mgr: &CancelManager,
) -> Result<String, String> {
    let dir = work_dir(&config.rect_id);
    fs::create_dir_all(&dir).map_err(|e| format!("create dir: {e}"))?;

    // Write existing code as context
    if let Some(code) = &config.existing_code {
        fs::write(dir.join("index.html"), code)
            .map_err(|e| format!("write context: {e}"))?;
    }

    // Build messages with revision history as previous user turns
    let mut messages: Vec<ChatMessage> = Vec::new();
    if let Some(history) = &config.history {
        for prompt in history {
            messages.push(ChatMessage {
                role: "user".to_string(),
                content: prompt.clone(),
            });
        }
    }
    messages.push(ChatMessage {
        role: "user".to_string(),
        content: config.prompt.clone(),
    });

    let cancel_rx = cancel_mgr.register(config.rect_id.clone());
    let response = client::chat_stream(settings, &messages, |_| {}, Some(cancel_rx)).await?;
    cancel_mgr.unregister(&config.rect_id);

    let code = extract_code(&response)
        .ok_or_else(|| "AI response did not contain a code block".to_string())?;

    // Write result so the tmp dir always has the output file (useful for debugging)
    fs::write(dir.join("index.html"), &code)
        .map_err(|e| format!("write result: {e}"))?;

    Ok(code)
}

// ── CLI-based agent (OpenCode, Claude Code, Custom) ─────────────────

/// Pre-create the work directory and index.html for a rect.
/// The agent is guided to modify this file rather than create new ones.
#[tauri::command]
pub async fn prepare_work_dir(rect_id: String, existing_code: Option<String>) -> Result<(), String> {
    let dir = work_dir(&rect_id);
    fs::create_dir_all(&dir).map_err(|e| format!("create dir: {e}"))?;
    let path = dir.join("index.html");
    if !path.exists() {
        fs::write(&path, existing_code.unwrap_or_default())
            .map_err(|e| format!("write index.html: {e}"))?;
    }
    Ok(())
}

async fn run_cli_agent(
    config: &AgentRunConfig,
    cancel_mgr: &CancelManager,
) -> Result<String, String> {
    let dir = work_dir(&config.rect_id);
    fs::create_dir_all(&dir).map_err(|e| format!("create dir: {e}"))?;

    let context_path = dir.join("index.html");

    // Always write index.html (existing code or empty) so the CLI agent
    // finds the file already there and modifies it instead of creating a new one.
    fs::write(&context_path, config.existing_code.as_deref().unwrap_or_default())
        .map_err(|e| format!("write context: {e}"))?;

    let dir_str = dir.to_string_lossy().to_string();
    let context_str = context_path.to_string_lossy().to_string();

    // Resolve command and arguments
    let command = config.command.as_deref().unwrap_or("opencode");
    let template = config.args_template.as_deref().unwrap_or("");
    let raw_args = interpolate(template, &dir_str, &context_str);

    // Build args from the template (whitespace-split for flags only),
    // then append the prompt as a single argument so multi-word prompts aren't broken.
    let mut args: Vec<&str> = raw_args.split_whitespace().collect();
    if !config.prompt.is_empty() {
        args.push(&config.prompt);
    }

    eprintln!("[agent] running: {command} {}", args.join(" "));
    eprintln!("[agent] cwd: {dir_str}");

    let mut child = tokio::process::Command::new(command)
        .args(&args)
        .current_dir(&dir)
        .stdout(std::process::Stdio::piped())
        .stderr(std::process::Stdio::piped())
        .spawn()
        .map_err(|e| format!("cannot spawn {command}: {e}"))?;

    let mut stdout_pipe = child.stdout.take();
    let mut stderr_pipe = child.stderr.take();

    /// Read bytes from a pipe into a string.
    async fn read_pipe_str<T: tokio::io::AsyncRead + Unpin>(pipe: &mut Option<T>) -> Option<String> {
        let mut buf = String::new();
        pipe.as_mut()?.read_to_string(&mut buf).await.ok()?;
        let trimmed = buf.trim().to_string();
        if trimmed.is_empty() { None } else { Some(trimmed) }
    }

    /// Helper to format a failure message with exit code + captured output.
    fn fail_msg(command: &str, status: std::process::ExitStatus, stdout: Option<String>, stderr: Option<String>) -> String {
        let code = status.code().map(|c| c.to_string()).unwrap_or_else(|| "unknown".into());
        let mut msg = format!("{command} exited with status {code}");
        if let Some(out) = &stdout { msg.push_str(&format!("\nstdout:\n{out}")); }
        if let Some(err) = &stderr { msg.push_str(&format!("\nstderr:\n{err}")); }
        msg
    }

    // Race with cancellation
    let mut cancel_rx = cancel_mgr.register(config.rect_id.clone());

    let status = tokio::select! {
        status = child.wait() => {
            cancel_mgr.unregister(&config.rect_id);
            status.map_err(|e| format!("{command} error: {e}"))?
        }
        _ = cancel_rx.changed() => {
            child.kill().await.ok();
            child.wait().await.ok();
            return Err("cancelled".into());
        }
    };

    if !status.success() {
        let stderr = read_pipe_str(&mut stderr_pipe).await;
        let stdout = read_pipe_str(&mut stdout_pipe).await;
        return Err(fail_msg(command, status, stdout, stderr));
    }

    // Always read index.html — the agent was guided to modify this file.
    let content = fs::read_to_string(&context_path)
        .map_err(|e| format!("cannot read index.html: {e}"))?;
    Ok(extract_code(&content).unwrap_or(content))
}

// ── Public command ──────────────────────────────────────────────────

#[tauri::command]
pub async fn run_agent_generate(
    config: AgentRunConfig,
    settings: AiSettings,
    cancel_mgr: tauri::State<'_, CancelManager>,
) -> Result<String, String> {
    match config.agent_type.as_str() {
        "streaming" => run_streaming_agent(&config, &settings, &cancel_mgr).await,
        "custom" => run_cli_agent(&config, &cancel_mgr).await,
        agent_type => {
            // For named agents (opencode, claude, etc.) — always use defaults,
            // ignore any custom command/template the frontend may have sent.
            let defaults = builtin_agent_config(agent_type)?;
            let merged = AgentRunConfig {
                command: defaults.command,
                args_template: defaults.args_template,
                ..config
            };
            run_cli_agent(&merged, &cancel_mgr).await
        }
    }
}

/// Detect if a CLI binary is available on PATH.
#[tauri::command]
pub fn check_agent_installed(command: String) -> bool {
    which::which(&command).is_ok()
}

// ── Built-in agent configs ──────────────────────────────────────────

struct AgentDefaults {
    command: Option<String>,
    args_template: Option<String>,
}

fn builtin_agent_config(agent_type: &str) -> Result<AgentDefaults, String> {
    match agent_type {
        "opencode" => Ok(AgentDefaults {
            command: Some("opencode".into()),
            args_template: Some(r#"run --dir {dir} --format json"#.into()),
        }),
        "claude" => Ok(AgentDefaults {
            command: Some("claude".into()),
            args_template: Some("-p".into()),
        }),
        other => Err(format!("unknown built-in agent type: {other}")),
    }
}

File: electron/ai/agent.js
import { ipcMain } from "electron";
import fs from "fs";
import path from "path";
import { homedir } from "os";
import { spawn } from "child_process";
import { chatWithTools } from "./client.js";
import { getCancelManager } from "../ipc/cancel.js";
import { execSync } from "child_process";

// ── Helpers ────────────────────────────────────────────────

function workDir(rectId) {
  const base = path.join(homedir(), ".pagesprite", "tmp", rectId);
  return base;
}

/** Extract code from AI response — tries ```html first, then any ``` block */
function extractCode(text) {
  const htmlMatch = text.match(/```html\n?([\s\S]*?)```/);
  if (htmlMatch) return htmlMatch[1].trim();
  const anyMatch = text.match(/```\w*\n?([\s\S]*?)```/);
  if (anyMatch) return anyMatch[1].trim();
  return null;
}

/** Interpolate {dir} and {contextFile} in a template string */
function interpolate(template, dir, contextFile) {
  return template.replace(/\{dir\}/g, dir).replace(/\{contextFile\}/g, contextFile);
}

/**
 * Build the complete prompt structure shared by all agent types:
 *   1. Fixed system prompt
 *   2. Style prompt (if set)
 *   3. Previous request history (if any)
 *   4. Current request
 * Returns { system, history, current } so each agent can format as appropriate.
 */
function buildPromptParts(config, settings, isStreaming) {
  // History prompts with labels
  const historyTexts = [];
  if (config.history) {
    for (let i = 0; i < config.history.length; i++) {
      historyTexts.push(`[Previous Request ${i + 1}]\n${config.history[i]}`);
    }
  }
  const base = isStreaming ? SYSTEM_PROMPT_STREAMING : SYSTEM_PROMPT_COMMON;
  return {
    system: base + (settings.system_prompt ? "\n\n" + settings.system_prompt : ""),
    history: historyTexts,
    current: config.prompt || "",
  };
}

/** Format the full prompt as a single text block (for CLI agents). */
function formatCliPrompt(parts) {
  const blocks = [parts.system, ...parts.history, parts.current];
  return blocks.filter(Boolean).join("\n\n");
}

// ── Tool definitions ────────────────────────────────────────

const TOOLS = [
  {
    type: "function",
    function: {
      name: "read",
      description: "Read a file from the working directory",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string", description: "File path relative to working directory" },
        },
        required: ["path"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "write",
      description: "Write content to a file in the working directory",
      parameters: {
        type: "object",
        properties: {
          path: { type: "string", description: "File path relative to working directory" },
          content: { type: "string", description: "File content" },
        },
        required: ["path", "content"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "list",
      description: "List all files in the working directory with their sizes",
      parameters: { type: "object", properties: {} },
    },
  },
  {
    type: "function",
    function: {
      name: "finish",
      description: "Call this when generation is complete and you are satisfied with the result. The content of index.html will be returned as the final output.",
      parameters: { type: "object", properties: {} },
    },
  },
];

const SYSTEM_PROMPT_COMMON = `You are a frontend page generator. Generate complete, standalone HTML files with inline CSS and JavaScript.

Write the main file to index.html.

Requirements:
- Output a single HTML file with embedded <style> and <script> tags
- Use modern CSS (flexbox, grid, custom properties)
- Make designs responsive and visually polished
- Use CSS transitions/animations sparingly for polish
- Do NOT use external CDN links (except for icon libraries if essential)
- **Font Awesome 6 Free is built-in. Use its icon classes** (e.g., <i class="fas fa-home"></i>) for any icons needed. No need to include any Font Awesome CSS — it is injected automatically.

When working on a revision, the existing code is already in your working directory's index.html file. Read it from there to understand the current state, then modify the file with your complete generated HTML output, replacing its contents entirely. Do NOT create any new files.

When the user provides annotations, they describe specific areas of the rendered page that need changes. Pay close attention to the annotation positions and text.`;

const SYSTEM_PROMPT_STREAMING = SYSTEM_PROMPT_COMMON + `

Workflow:
- Use the read/write/list tools to review and refine your output
- Iterate as many times as needed
- Only call finish() when you are truly satisfied with the result.`;

/** Execute a tool call. Returns the result string to send back to the model. */
function executeTool(name, args, dir) {
  switch (name) {
    case "read": {
      if (!args.path) return `Error: Missing required argument "path" for read`;
      const p = path.join(dir, args.path);
      if (!fs.existsSync(p)) return `Error: File not found: ${args.path}`;
      return fs.readFileSync(p, "utf-8");
    }
    case "write": {
      if (!args.path) return `Error: Missing required argument "path" for write`;
      const p = path.join(dir, args.path);
      fs.mkdirSync(path.dirname(p), { recursive: true });
      fs.writeFileSync(p, args.content || "");
      return `OK: Written ${args.path} (${(args.content || "").length} chars)`;
    }
    case "list": {
      const items = fs.readdirSync(dir);
      if (items.length === 0) return "(empty directory)";
      return items.map((name) => {
        const stat = fs.statSync(path.join(dir, name));
        return `${name} (${stat.size} bytes)`;
      }).join("\n");
    }
    default:
      return `Error: Unknown tool: ${name}`;
  }
}

// ── Streaming agent ────────────────────────────────────────

async function runToolUseAgent(config, settings, cancelMgr) {
  const dir = workDir(config.rect_id);
  fs.mkdirSync(dir, { recursive: true });

  if (config.existing_code) {
    fs.writeFileSync(path.join(dir, "index.html"), config.existing_code);
  }

  // Build unified prompt structure
  const promptParts = buildPromptParts(config, settings, true);
  const messages = [
    ...promptParts.history.map((content) => ({ role: "user", content })),
    { role: "user", content: promptParts.current },
  ];
  const toolSettings = { ...settings, system_prompt: promptParts.system };

  console.error("[agent] system prompt:", toolSettings.system_prompt);
  console.error("[agent] user messages count:", messages.length);
  messages.forEach((m, i) => console.error("[agent] msg " + i + ":", m.content));

  const MAX_ITERATIONS = 25;
  const controller = cancelMgr.registerAbort(config.rect_id);

  try {
    for (let i = 0; i < MAX_ITERATIONS; i++) {
      // eslint-disable-next-line no-await-in-loop
      const reply = await chatWithTools(toolSettings, messages, TOOLS, controller.signal);

      // Add assistant message to conversation history
      messages.push({
        role: "assistant",
        content: reply.content || null,
        tool_calls: reply.tool_calls,
      });

      // No tool calls → model is responding with text. Try to finish.
      if (!reply.tool_calls || reply.tool_calls.length === 0) {
        const code = extractCode(reply.content || "");
        if (code) fs.writeFileSync(path.join(dir, "index.html"), code);
        return fs.readFileSync(path.join(dir, "index.html"), "utf-8");
      }

      // Process tool calls
      for (const tc of reply.tool_calls) {
        if (tc.type !== "function") continue;
        const { name, arguments: argsStr } = tc.function;

        // finish() → generation complete
        if (name === "finish") {
          cancelMgr.unregisterAbort(config.rect_id);
          return fs.readFileSync(path.join(dir, "index.html"), "utf-8");
        }

        // Parse arguments and execute
        let args;
        try { args = JSON.parse(argsStr); } catch { args = {}; }
        const result = executeTool(name, args, dir);

        messages.push({
          role: "tool",
          tool_call_id: tc.id,
          content: result,
        });
      }
    }

    // Max iterations reached — return whatever is on disk
    cancelMgr.unregisterAbort(config.rect_id);
    const fallbackPath = path.join(dir, "index.html");
    if (fs.existsSync(fallbackPath)) {
      const html = fs.readFileSync(fallbackPath, "utf-8");
      if (html.trim()) return html;
    }
    throw new Error("Agent did not produce a result within the maximum number of iterations");
  } catch (err) {
    cancelMgr.unregisterAbort(config.rect_id);
    if (err.name === "AbortError") throw new Error("cancelled");
    throw err;
  }
}

// ── CLI agent ──────────────────────────────────────────────

async function runCliAgent(config, cancelMgr, webContents) {
  const dir = workDir(config.rect_id);
  fs.mkdirSync(dir, { recursive: true });

  const contextPath = path.join(dir, "index.html");
  fs.writeFileSync(contextPath, config.existing_code || "");

  const command = config.command || "opencode";
  const template = config.args_template || "";
  const rawArgs = interpolate(template, dir, contextPath);

  const args = rawArgs ? rawArgs.split(/\s+/) : [];
  if (config.prompt) args.push(config.prompt);

  console.error(`[agent] running: ${command} ${args.join(" ")}`);

  const child = spawn(command, args, {
    cwd: dir,
    stdio: ["ignore", "pipe", "pipe"],
  });

  cancelMgr.registerChild(config.rect_id, child);

  // Collect stdout/stderr and stream progress to renderer
  let stdout = "";
  let stderr = "";
  child.stdout.on("data", (d) => {
    const text = d.toString();
    stdout += text;
    if (webContents && !webContents.isDestroyed()) {
      webContents.send("agent:progress", { rectId: config.rect_id, type: "stdout", text });
    }
  });
  child.stderr.on("data", (d) => {
    const text = d.toString();
    stderr += text;
    if (webContents && !webContents.isDestroyed()) {
      webContents.send("agent:progress", { rectId: config.rect_id, type: "stderr", text });
    }
  });

  try {
    const code = await new Promise((resolve, reject) => {
      child.on("close", (exitCode) => {
        cancelMgr.unregisterChild(config.rect_id);
        if (exitCode === 0) {
          const content = fs.readFileSync(contextPath, "utf-8");
          resolve(extractCode(content) || content);
        } else {
          let msg = `${command} exited with status ${exitCode}`;
          if (stdout.trim()) msg += `\nstdout:\n${stdout.trim()}`;
          if (stderr.trim()) msg += `\nstderr:\n${stderr.trim()}`;
          reject(new Error(msg));
        }
      });
      child.on("error", (err) => {
        cancelMgr.unregisterChild(config.rect_id);
        reject(new Error(`${command} error: ${err.message}`));
      });
    });
    return code;
  } catch (err) {
    if (child.killed) throw new Error("cancelled");
    cancelMgr.unregisterChild(config.rect_id);
    throw err;
  }
}

// ── Built-in agent configs ─────────────────────────────────

const BUILTIN_CONFIGS = {
  opencode: { command: "opencode", args_template: "run --dir {dir} --format json" },
  claude: { command: "claude", args_template: "-p" },
};

// ── IPC handlers ───────────────────────────────────────────

export function registerAgentHandlers(webContents) {
  const cancelMgr = getCancelManager();

  ipcMain.handle("agent:generate", async (_, config, settings) => {
    switch (config.agent_type) {
      case "streaming":
        return runToolUseAgent(config, settings, cancelMgr);
      case "custom": {
        const promptParts = buildPromptParts(config, settings, false);
        const fullPrompt = formatCliPrompt(promptParts);
        console.error("[agent] cli prompt:", fullPrompt);
        return runCliAgent({ ...config, prompt: fullPrompt }, cancelMgr, webContents);
      }
      default: {
        // Named agents (opencode, claude, etc.) — use built-in defaults
        const defaults = BUILTIN_CONFIGS[config.agent_type];
        if (!defaults) throw new Error(`unknown agent type: ${config.agent_type}`);
        // Only carry forward rect_id/prompt/existing_code/history from the
        // renderer — command and args_template MUST come from BUILTIN_CONFIGS
        // so a stale agent_command in user settings cannot override them.
        const promptParts = buildPromptParts(config, settings, false);
        const fullPrompt = formatCliPrompt(promptParts);
        console.error("[agent] cli prompt:", fullPrompt);
        const merged = {
          ...defaults,
          rect_id: config.rect_id,
          prompt: fullPrompt,
          existing_code: config.existing_code,
          history: config.history,
        };
        return runCliAgent(merged, cancelMgr, webContents);
      }
    }
  });

  ipcMain.handle("agent:prepareWorkDir", async (_, rectId, existingCode) => {
    const dir = workDir(rectId);
    fs.mkdirSync(dir, { recursive: true });
    const p = path.join(dir, "index.html");
    if (!fs.existsSync(p)) {
      fs.writeFileSync(p, existingCode || "");
    }
  });

  ipcMain.handle("agent:checkInstalled", async (_, command) => {
    try {
      const cmd = process.platform === "win32" ? "where" : "which";
      execSync(`${cmd} ${command}`, { stdio: "ignore" });
      return true;
    } catch {
      return false;
    }
  });
}
